lib/backend/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,263 @@
+	}
+	return nil, trace.Wrap(err)
+}
+
+// readSecretData reads the secret content and extracts the content for key.
+// returns an error if the key does not exist or the data is empty.
+func (b *Backend) readSecretData(ctx context.Context, key []byte) (*backend.Item, error) {
+	secret, err := b.getSecret(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	data, ok := secret.Data[backendKeyToSecret(key)]
+	if !ok || len(data) == 0 {
+		return nil, trace.NotFound("key [%s] not found in secret %s", string(key), b.secretName)
+	}
+
+	return &backend.Item{
+		Key:   key,
+		Value: data,
+	}, nil
+}
+
+func (b *Backend) updateSecretContent(ctx context.Context, items ...backend.Item) (*backend.Lease, error) {
+	// FIXME(tigrato):
+	// for now, the agent is the owner of the secret so it's safe to replace changes
+
+	secret, err := b.getSecret(ctx)
+	if err != nil && trace.IsNotFound(err) {
+		_, err = b.createSecret(ctx, items...)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return &backend.Lease{}, nil
+	} else if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if secret.Data == nil {
+		secret.Data = map[string][]byte{}
+	}
+
+	updateDataMap(secret.Data, items...)
+
+	if err := b.updateSecret(ctx, secret); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &backend.Lease{}, nil
+}
+
+func (b *Backend) updateSecret(ctx context.Context, secret *corev1.Secret) error {
+	secretApply := applyconfigv1.Secret(b.secretName, b.namespace).
+		WithResourceVersion(secret.ResourceVersion).
+		WithData(secret.Data).
+		WithLabels(secret.GetLabels()).
+		WithAnnotations(secret.GetAnnotations())
+
+	_, err := b.k8sClientSet.
+		CoreV1().
+		Secrets(b.namespace).
+		Apply(ctx, secretApply, metav1.ApplyOptions{FieldManager: b.replicaName})
+
+	return trace.Wrap(err)
+}
+
+func (b *Backend) createSecret(ctx context.Context, items ...backend.Item) (*corev1.Secret, error) {
+	const (
+		helmReleaseNameAnnotation     = "meta.helm.sh/release-name"
+		helmReleaseNamesaceAnnotation = "meta.helm.sh/release-namespace"
+		helmK8SManaged                = "app.kubernetes.io/managed-by"
+		helmResourcePolicy            = "helm.sh/resource-policy"
+	)
+	data := map[string][]byte{}
+	updateDataMap(data, items...)
+
+	secretApply := applyconfigv1.Secret(b.secretName, b.namespace).
+		WithData(data).
+		WithLabels(map[string]string{
+			helmK8SManaged: "Helm",
+		}).
+		WithAnnotations(map[string]string{
+			helmReleaseNameAnnotation:     os.Getenv(releaseNameEnv),
+			helmReleaseNamesaceAnnotation: os.Getenv(namespaceEnv),
+			helmResourcePolicy:            "keep",
+		})
+
+	return b.k8sClientSet.
+		CoreV1().
+		Secrets(b.namespace).
+		Apply(
+			ctx,
+			secretApply,
+			metav1.ApplyOptions{FieldManager: b.replicaName},
+		)
+
+}
+
+// backendKeyToSecret replaces the "/" with "."
+// "/" chars are not allowed in Kubernetes Secret keys.
+func backendKeyToSecret(k []byte) string {
+	return strings.ReplaceAll(string(k), "/", ".")
+}
```

Comment:
Just to be defensive against future changes, should we also decide how to handle dots in the names? It shouldn't ever happen with identities and state in process storage, but still.

Reply:
We can support `/` by marshaling the secret data to json and moving it to a `data` key. 
```
data:
   "data": "{ "key1/data/":data1..,"key2/data":...}"
```